src/realm/sync/noinst/sync_metadata_schema.cpp

```diff
@@ -0,0 +1,220 @@
+        }
+
+        for (auto& col : table.columns) {
+            auto col_key = table_ref->get_column_key(col.name);
+            if (!col_key) {
+                throw std::runtime_error(
+                    util::format("column %1 is missing in sync internal table %2", col.name, table.name));
+            }
+
+            auto found_col_type = table_ref->get_column_type(col_key);
+            if (found_col_type != col.data_type) {
+                throw std::runtime_error(
+                    util::format("column %1 in sync internal table %2 is the wrong type", col.name, table.name));
+            }
+
+            if (col.is_optional != table_ref->is_nullable(col_key)) {
+                throw std::runtime_error(
+                    util::format("column %1 in sync internal table %2 has different nullabilty than in its schema",
+                                 col.name, table.name));
+            }
+
+            if (col.data_type == type_LinkList &&
+                table_ref->get_link_target(col_key)->get_name() != col.target_table) {
+                throw std::runtime_error(
+                    util::format("column %1 in sync internal table %2 links to the wrong table %3", col.name,
+                                 table.name, table_ref->get_link_target(col_key)));
+            }
+            *col.key_out = col_key;
+        }
+    }
+}
+
+SyncMetadataSchemaVersions::SyncMetadataSchemaVersions(const TransactionRef& tr)
+{
+    TableKey legacy_table_key;
+    ColKey legacy_version_key;
+    std::vector<SyncMetadataTable> legacy_table_def{
+        {&legacy_table_key, c_flx_metadata_table, {{&legacy_version_key, c_meta_schema_version_field, type_Int}}}};
+    std::vector<SyncMetadataTable> unified_schema_version_table_def{
+        {&m_table,
+         c_sync_internal_schemas_table,
+         {&m_table_group_field, c_meta_schema_table_group_field, type_String},
+         {{&m_version_field, c_meta_schema_version_field, type_Int}}}};
+
+    if (!m_table) {
+        if (tr->has_table(c_sync_internal_schemas_table)) {
+            load_sync_metadata_schema(tr, &unified_schema_version_table_def);
+        }
+        else {
+            tr->promote_to_write();
+            create_sync_metadata_schema(tr, &unified_schema_version_table_def);
+            tr->commit_and_continue_as_read();
+        }
+    }
+
+    if (!tr->has_table(c_flx_metadata_table)) {
+        return;
+    }
+
+    load_sync_metadata_schema(tr, &legacy_table_def);
+    // Migrate from just having a subscription store metadata table to having multiple table groups with multiple
+    // versions.
+    tr->promote_to_write();
+    auto legacy_meta_table = tr->get_table(legacy_table_key);
+    auto legacy_obj = legacy_meta_table->get_object(0);
+    set_version_for(tr, internal_table_groups::c_flx_subscription_store, legacy_obj.get<int64_t>(legacy_version_key));
+    tr->remove_table(legacy_table_key);
+    tr->commit_and_continue_as_read();
```

Comment:
isn't commit enough here?

Reply:
Since we're taking a transaction ref, we don't know if the caller is going to use the transaction for something else after this constructor finishes. So I decided to leave the transaction in the state we got it in.